distributed/s3fs.py

```diff
@@ -0,0 +1,539 @@
+            self.cache = self.cache + new
+
+    def read(self, length=-1):
+        """
+        Return data from cache, or fetch pieces as necessary
+        """
+        if self.mode != 'rb':
+            raise ValueError('File not in read mode')
+        if length < 0:
+            length = self.size
+        if self.closed:
+            raise ValueError('I/O operation on closed file.')
+        self._fetch(self.loc, self.loc + length)
+        out = self.cache[self.loc - self.start:
+                         self.loc - self.start + length]
+        self.loc += len(out)
+        return out
+
+    def write(self, data):
+        """
+        Write data to buffer.
+
+        Buffer only sent to S3 on flush().
+        """
+        if self.mode != 'wb':
+            raise ValueError('File not in write mode')
+        if self.closed:
+            raise ValueError('I/O operation on closed file.')
+        return self.buffer.write(data)
+
+    def flush(self):
+        """
+        Write buffered data to S3.
+        """
+        if self.mode == 'wb':
+            try:
+                self.s3.s3.head_bucket(Bucket=self.bucket)
+            except ClientError:
+                try:
+                    self.s3.s3.create_bucket(Bucket=self.bucket)
+                except ClientError:
+                    raise IOError('Create bucket failed: %s', self.bucket)
+            pos = self.buffer.tell()
+            self.buffer.seek(0)
+            try:
+                out = self.s3.s3.put_object(Bucket=self.bucket, Key=self.key,
+                                            Body=self.buffer.read())
+            finally:
+                self.buffer.seek(pos)
+            self.s3._ls(self.bucket, refresh=True)
+            if out['ResponseMetadata']['HTTPStatusCode'] != 200:
+                raise IOError("Write failed: %s", out)
+
+    def close(self):
+        self.flush()
+        self.cache = None
+        self.closed = True
+
+    def __del__(self):
+        self.close()
+
+    def __str__(self):
+        return "<S3File %s/%s>" % (self.bucket, self.key)
+
+    __repr__ = __str__
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self.close()
+
+MAX_ATTEMPTS = 10
+
+
+def _fetch_range(client, bucket, key, start, end):
+    try:
+        for i in range(MAX_ATTEMPTS):
+            try:
+                resp = client.get_object(Bucket=bucket, Key=key,
+                                         Range='bytes=%i-%i' % (start, end - 1))
+            except ClientError as e:
+                if e.response['Error'].get('Code', 'Unknown') in ['416', 'InvalidRange']:
+                    return b''
+            except Exception as e:
+                logger.debug('Exception %e on S3 download', e)
+                continue
+            buff = io.BytesIO()
+            buffer_size = 1024 * 16
+            for chunk in iter(lambda: resp['Body'].read(buffer_size),
+                              b''):
```

Comment:
This would happen if we raised a `ClientError` but didn't fall into the listed condition.